airflow-core/src/airflow/utils/otel_config.py

```diff
@@ -0,0 +1,181 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+from __future__ import annotations
+
+import logging
+import os
+from dataclasses import dataclass
+from enum import Enum
+from functools import lru_cache
+
+log = logging.getLogger(__name__)
+
+
+def _parse_kv_str_to_dict(str_var: str) -> dict[str, str]:
+    """
+    Convert a string of key-value pairs to a dictionary.
+
+    Environment variables like 'OTEL_RESOURCE_ATTRIBUTES' or 'OTEL_EXPORTER_OTLP_HEADERS'
+    accept values with the format "key1=value1,key2=value2,..."
+    """
+    configs = {}
+    if str_var:
+        for pair in str_var.split(","):
+            if "=" in pair:
+                k, v = pair.split("=", 1)
+                configs[k.strip()] = v.strip()
+    return configs
+
+
+class OtelDataType(str, Enum):
+    """Enum with the different telemetry data types."""
+
+    TRACES = "traces"
+    METRICS = "metrics"
+    LOGS = "logs"
+
+
+@dataclass(frozen=True)
+class OtelConfig:
+    """Immutable class for holding and validating OTel config environment variables."""
+
+    data_type: OtelDataType  # traces | metrics
+    endpoint: str  # url
+    protocol: str  # "grpc" or "http/protobuf"
+    exporter: str  # OTEL_TRACES_EXPORTER | OTEL_METRICS_EXPORTER
+    service_name: str  # default "Airflow"
+    headers_kv_str: str
+    headers: dict[str, str]
+    resource_attributes_kv_str: str
+    resource_attributes: dict[str, str]
+    interval_ms: int
+
+    def __post_init__(self):
+        """Validate the environment variables where necessary."""
+        endpoint_type_specific = (
+            "OTEL_EXPORTER_OTLP_TRACES_ENDPOINT"
+            if self.data_type == OtelDataType.TRACES
+            else "OTEL_EXPORTER_OTLP_METRICS_ENDPOINT"
+        )
+
+        if not self.endpoint:
+            raise OSError(
```

Comment:
Would `ValueError` be better ?
```suggestion
            raise ValueError(
```

Reply:
I don't have a personal preference. I changed it.